### PR TITLE
Initial integration with webpack encore

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this breaks bc in many ways.

## Changelog

```markdown
### Added
- Added integration with webpack encore
- Npm as a package manager for internal assets

### Changed
- Changed paths of javascript and css files

### Removed
- Bower as a package manager
```

There are some libraries that currently cannot be integrated with npm:

- eonasdan-bootstrap-datetimepicker (in its current version)
- jquery.confirmExit (it wasn't even integrated with bower, it does not exists on npm)
- jqueryui

I have moved the SonataCore files to SonataAdmin Resources too.

This is the initial step of the integration, there is so much work to do, but this allows us to get rid of almost 200k lines of code that does not belong to this repository.

Next steps... when this pr is merged:

- Integrate / Remove missing dependencies
- Upgrade some dependencies (and stop fixing versions on package.json)
- Remove ionicons
- Integrate our code with the webpack encore process
- Documentation about the frontend assets for contributors
- Documentation about the frontend assets for users of the bundle
- Consider integration with webpack encore bundle
- Generate the dist files only on release (maybe not necessary)
- Add stylelint and eslint
- Remove x-editable (or make it compatible with boostrap 4 or implement own solution)
- Upgrade admin lte to v3 (check sandbox branch and older prs on sonata admin)

The thing is that we can't afford to do all this at once, so we better have something than nothing.

### CODE REVIEW

**Comment by jordisala1991** [comment]
Probably I am missing some upgrade notes and tests removal for the corebundle assets thing, but wdyt as a first step ? @sonata-project/contributors

File: docs/reference/configuration.rst
Comment (core23) [inline docs/reference/configuration.rst:179]: Would appreciate if we can completely remove this part and replace this with one webpack_encore entrypoint.

```yaml
webpack_encore:
    builds:
        sonata_admin: '%kernel.project_dir%/public/bundles/sonataadmin

```

Comment (jordisala1991) [reply docs/reference/configuration.rst:179]: IMHO this PR is big enough, and I plan to do it at some point but please, read the next steps section, there is a huge list of things to do and consider

**Review by phansys — DISMISSED** [review]

File: docs/reference/configuration.rst
Comment (jordisala1991) [reply docs/reference/configuration.rst:179]: Just added your consideration. 

Also to be note, I took inspiration of how easy admin handles its js and css, and I didnt found the webpack encore bundle integration in place, they just generate the minifies files but include them normally.

Maybe @javiereguiluz could help us with why he took that path.

Anyway , as I said, for me this should be considered after this Pr to keep them small enough.

**Comment by phansys** [comment]
How will we handle the required upgrade notes for these changes?

**Comment by jordisala1991** [comment]
> How will we handle the required upgrade notes for these changes?

I was planning to add a section mentioning the changes on css and build process, with a note about:

Removal of Bower
Added npm.js
Modification of some assets location

Also we should probably update the mentions of bower install and build process for the npm and encore process

wdyt?

**Comment by wbloszyk** [comment]
@jordisala1991 
Can you look at `webpack` branch in `sandbox` too?

**Comment by jordisala1991** [comment]
> @jordisala1991 
> 
> Can you look at `webpack` branch in `sandbox` too? 

Yes, of course, added to the next steps sections. IMO we should take a look at all those prs and branches when dealing with the upgrade of admin lte. Also, since I plan to split this huge effort as much as possible, it can be done in multiple Prs and multiple people not only me.

File: docs/reference/configuration.rst
Comment (javiereguiluz) [reply docs/reference/configuration.rst:179]: @jordisala1991 I don't remember exactly why we did this 🤔  But yes, we may need to include our own styles using just Webpack Encore entries (and leave the option to include raw CSS files for those who need it). I'll add this to my TODO list. Cheers!

**Review by wbloszyk — DISMISSED** [review]

**Comment by wbloszyk** [comment]
Looks good for me. In monday I will update issues for webpack and bootstrap update for whole Sonata bundles. Happy weekend for every one.

**Comment by jordisala1991** [comment]
Added upgrade note @phansys

File: UPGRADE-4.0.md
Comment (phansys) [inline UPGRADE-4.0.md:10]:
```suggestion
Frontend dependencies are handled with NPM. Bower is not used anymore.
```

File: UPGRADE-4.0.md
Comment (phansys) [inline UPGRADE-4.0.md:12]:
```suggestion
A lot of assets that were previously public are handled with NPM and placed in a private `node_modules/` directory. 
From these dependencies, only the necessary files are exposed publicly through Webpack Encore.
```

File: UPGRADE-4.0.md
Comment (phansys) [inline UPGRADE-4.0.md:14]:
```suggestion
Please check the `src/Resources/public/dist` and the documentation to see the used CSS, JavaScript, images and fonts.
```

File: UPGRADE-4.0.md
Comment (phansys) [inline UPGRADE-4.0.md:16]:
```suggestion
If you are customising (specially removing standard JavaScript or CSS) the assets, this will affect you.
```

**Review by phansys — CHANGES_REQUESTED** [review]
I left some suggestions about the upgrade notes, having 2 premises in mind:
* These instructions are provided from SonataAdminBundle (and for SonataAdminBundle), so there is no need to mention the package name in the comments;
* The focus near "our" / "you" referencing "the package" / "users of the package" was replaced to bring descriptions centered in the required changes the consumers of this package must consider when upgrading.

**Review by phansys — DISMISSED** [review]

**Comment by franmomu** [comment]
naive question: do we have to execute `yarn` in the building process?

**Comment by jordisala1991** [comment]
> naive question: do we have to execute `yarn` in the building process?

Yes, to have the final `src/Resources/public/dist` i have executed:

```
yarn install (but once we have the lock file, the command equivalent to composer install is: 'yarn install --frozen-lockfile')
yarn encore production
```

On another PR we could document how to do this (and probably add some commands to the makefile). I will add this to the list of next steps

BTW I will convert this list to an issue, once this one is merged.

**Comment by franmomu** [comment]
Thanks! I wasn't clear, I meant to us as contributors. 

> On another PR we could document how to do this (and probably add some commands to the makefile). I will add this to the list of next steps

That's perfect, the persistence bundles execute some tests using `symfony/panther` and need `JavaScript`, so I guess we should execute those commands in the Github actions.

**Review by franmomu — APPROVED** [review]

**Comment by jordisala1991** [comment]
> Thanks! I wasn't clear, I meant to us as contributors. 
> 
> 
> 
> > On another PR we could document how to do this (and probably add some commands to the makefile). I will add this to the list of next steps
> 
> 
> 
> That's perfect, the persistence bundles execute some tests using `symfony/panther` and need `JavaScript`, so I guess we should execute those commands in the Github actions.
> 
> 
> 
> 

If those test are on the persistence bundle, the admin bundle will have the requires js already compiled. For now you wont need to execute anything more than before

**Comment by VincentLanglet** [comment]
If we merge this, does it block the (alpha or not) release of 4.0 ?
Should we open an issue for the TO-DO ?

**Review by VincentLanglet — CHANGES_REQUESTED** [review]
I'd like to release the alpha before this merge

**Comment by jordisala1991** [comment]
> If we merge this, does it block the (alpha or not) release of 4.0 ?
> Should we open an issue for the TO-DO ?

I will open an issue with the TO-DO of course. But having that to-do does not mean that this PR alone does not bring value. I would prefer going with only this PR rather than keeping bower TBH

> I'd like to release the alpha before this merge

We can have multiple alphas too. If the plan is to tag it soon, I'm okay not having on the first one. But if the plan is to wait more time, I would prefer this merged, because I am already working on next PR's

**Comment by VincentLanglet** [comment]
> > I'd like to release the alpha before this merge
> 
> We can have multiple alphas too. If the plan is to tag it soon, I'm okay not having on the first one. But if the plan is to wait more time, I would prefer this merged, because I am already working on next PR's

The plan is to release the alpha today.

**Comment by jordisala1991** [comment]
> > > I'd like to release the alpha before this merge
> > 
> > 
> > We can have multiple alphas too. If the plan is to tag it soon, I'm okay not having on the first one. But if the plan is to wait more time, I would prefer this merged, because I am already working on next PR's
> 
> The plan is to release the alpha today.

Oh, then better tag the alpha-1 without this. And we can merge after

**Review by franmomu — APPROVED** [review]

**Review by franmomu — APPROVED** [review]

**Review by core23 — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Pull request merged**
